### PR TITLE
fix: snapshot test failures on some systems

### LINKED ISSUE #3392: failed snapshot tests on some systems
*Test File*
`test/manager/pip_setup/extract.spec.js`

*Function Failing Test*
`extractSetupFile`

*Failed Test*
`Snapshot Test`

*Reason*
The function is calling `python` command which can be aliased to different python versions on different systems, the function depends on `mock` library which comes pre-installed on python v3.7+

*Proposed solution*
Get the python alias which points to v3.7+ before calling the `python` command and call that alias instead of `python`

https://www.useloom.com/share/3539335673b341f2b0649ef6205da156

### PR DESCRIPTION
**Pull request by rishabh1403**
- The function is calling python command which can be aliased to different python versions on different systems, the function depends on mock library which comes pre-installed on python v3.7+
- Add a python alias detection command which gets the correct python command which is aliased to python3.7+

Closes #3392

### CODE REVIEW

File: lib/manager/pip_setup/extract.js [hunk 1]
```diff
@@ -8,6 +8,28 @@ module.exports = {
   extractPackageFile,
   extractSetupFile,
 };
+const cacheNamespace = 'python-alias';
```
Comment (rarkins) [inline lib/manager/pip_setup/extract.js]: Maybe move this closer to where it's used?

File: lib/manager/pip_setup/extract.js [hunk 2]
```diff
@@ -40,7 +62,13 @@ async function extractSetupFile(content, packageFile, config) {
     );
   } else {
     logger.info('Running python via global command');
-    cmd = 'python';
+    const alias = await renovateCache.get(cacheNamespace, 'pythonAlias');
+    if (alias) {
+      cmd = alias;
+    } else {
+      cmd = await getPythonAlias();
+      await renovateCache.set(cacheNamespace, 'pythonAlias', cmd);
```
Comment (rarkins) [inline lib/manager/pip_setup/extract.js]: I think we can cache this for a long time, perhaps a day? Default ttl if unspecified is 5 minutes.

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Actually what if instead of caching it stores it in its own state for the entire global run? Ie check and store it once  per invocation of Renovate

**Comment by rishabh1403** [comment]
> Actually what if instead of caching it stores it in its own state for the entire global run? Ie check and store it once per invocation of Renovate

That sounds good, but that would be helpful if this function is being called multiple times per invocation of renovate, is that the case?

**Comment by rarkins** [comment]
I think it’s called every time you need to parse a pip setup.py file

**Comment by rishabh1403** [comment]
@rarkins Please review

File: lib/manager/pip_setup/extract.js [hunk 3]
```diff
@@ -4,11 +4,41 @@ const { join } = require('upath');
...
+  for (const pythonVersion of pythonVersions) {
+    try {
+      const { stdout, stderr } = await exec(`${pythonVersion} --version`);
+      const version = parsePythonVersion(stdout || stderr);
+      // istanbul ignore if
+      if (version[0] >= 3 && version[1] >= 7) {
+        pythonAlias = pythonVersion;
+      }
+    } catch (err) {
+      logger.info(`${pythonVersion} alias not found`);
```
Comment (rarkins) [inline lib/manager/pip_setup/extract.js]: Logger.debug not logger.info. Assuming it’s quite common to have python 2 as default then there’s no need to pollute the info log or confuse users

**Review by rarkins — APPROVED** [review]

**Pull request merged**